readme.md
Marquee
---------------

[Marquee] is yet another web framework built on Mojo toolkit in
[Mojolicious] distribution, designed to make dynamic website
development to be plotted at an extension of traditional designer work.

This may possibly be a reinvention of [Mason], [Markup::Perl], [HTML::Embperl]
or PHP.

__This software is considered to be alpha quality and isn't recommended for
regular usage.__

![Komodo Edit](http://mrqe.jamadam.com/screenshot/komodo.png "Komodo Edit")

### Default URL mapping

By default, [Marquee] automatically maps request paths to corresponding
file directory structure. This is similar to typical HTTP servers such as Apache
and this is the leading motivation of this project.
To keep the URI semantics corresponds to directory tree makes things simple.

Given the request path
    
    /news/sports/hockey.html

[Marquee] searches for the following templates or static files under
application root.

    /public_html/news/sports/hockey.html
    /public_html/news/sports/hockey.html.ep
    /public_html/news/sports/hockey.html.epl

The extensioning rule is same as [Mojolicious]. The second extension indecates
a handler for template rendering. The rule allows Marquee to detect Content-Type
and implicitly serve the header. This system is more resonable
(by comparison to PHP).

ep and epl are always available and you can also add your own handler easily.
You can also override the mapping rule with [Marquee::Plugin::Router] plugin bundled
in core.

### Perlish template

[Marquee] provedes [Mojo::Template] based template handler which allows templates
to be written in more Perlish instead of template specific syntax,
therefore, it requires less lerning cost (by comparison to [Mason]).

### Easy to install

[Marquee] is written in pure-perl and only depends on [Mojolicious] distribution
which is also a pure-perl, so you can even deploy them via FTP.
Though [Mojolicious] depends on perl-5.10.1 or higher, there is still an option
to adopt backport project [mojo-legacy] to run on perl-5.8.7.

### Mojo toolkit is available

Since [Marquee] is based on mojo, many mojo classes helps you on manipulating
such as HTTP requests, responses, DOM and JSON.

### Command line interface

In addition to Perl OOP API, Marquee also provides command line interface
to serve current directory contents as a web site, using [Mojo::Daemon].
This is useful for development environment or even trivial file sharing.
The API provides some useful option such as auto index, POD viewer,
Markdown viewer. You don't need any Apache things anymore.

## INSTALLATION

To install this module, run the following commands:

    $ curl http://mrqe-get.jamadam.com/ | sh

## DOCUMENTATIONS

For more information refer to following documentations.

- [Marquee::Guides::Cookbook](http://mrqe.jamadam.com/perldoc/Marquee/Guides/Cookbook) (Cookbook)
- [Marquee](http://mrqe.jamadam.com/perldoc/Marquee) (Marquee base class)
- [Index of Modules](http://mrqe.jamadam.com/perldoc/)
- [Marquee::Guides::Cookbook](http://mrqe.jamadam.com/perldoc/Marquee/Guides/Cookbook#COMMAND_LINE_INTERFACE) (Command line interface)

## SCREENSHOTS

Here is some screenshots of how Marquee look like.

### debug screen

![debug screen](http://mrqe.jamadam.com/screenshot/debug_screen.png "Debug screen")

### Auto index

![auto index](http://mrqe.jamadam.com/screenshot/autoindex.png "Auto Index")

### Auto tree

![auto tree](http://mrqe.jamadam.com/screenshot/autoindextree.png "Auto Index")

## REPOSITORY

[Marquee](https://github.com/jamadam/Marquee)

## CREDIT

Icons by [Yusuke Kamiyamane].

## COPYRIGHT AND LICENSE

Copyright (c) 2012 [jamadam]

This program is free software; you can redistribute it and/or
modify it under the [same terms as Perl itself].

[Marquee]:http://mrqe.jamadam.com/perldoc/Marquee
[Marquee::Plugin::Router]:http://mrqe.jamadam.com/perldoc/Marquee/Plugin/Router
[Mojolicious]:http://mojolicio.us/
[Mason]:http://search.cpan.org/~jswartz/Mason-2.20/lib/Mason.pm
[mojo-legacy]:https://github.com/jamadam/mojo-legacy
[Mojo::Template]:/perldoc/Mojo/Template
[Mojo::Daemon]:/perldoc/Mojo/Daemon
[same terms as Perl itself]:http://dev.perl.org/licenses/
[Yusuke Kamiyamane]:http://p.yusukekamiyamane.com/
[jamadam]: http://blog2.jamadam.com/
[Markup::Perl]:http://search.cpan.org/~mmathews/Markup-Perl-0.5/lib/Markup/Perl.pm
[HTML::Embperl]:http://search.cpan.org/~grichter/HTML-Embperl-1.3.6/Embperl.pod


Changes
Revision history for Marquee

0.42 2019/11/18
  - Fixed a bug where config method didn't work.
  - Fixed examples.

0.41 2019/04/12
  - Fix tests for Mojolicious 7.93 and later

0.40 2018/05/21
  - Fix for Mojolicious 7.80

0.39 2018/03/30
  - Fixes in favor of Mojolicious 7.71

0.38 2017/12/25
  - Updated dependent version of Mojolicious.
  - Confirmed change on content-length behavior on Mojolicious.
  - Ported home method of Mojo class.

0.37 2016/09/15
  - Fixed tests

0.36 2016/09/15
  - Added CLI delay option to emulate distant server.

0.35 2016/05/16
  - Fixed test

0.34 2016/03/30
  - Updated dependent version of perl to 6.57
  - Removed obsolete tests.

0.33 2015/11/24
  - Updated dependent version of perl to 5.20
  - Tested on perl-5.22

0.32 2015/09/01
  - Relocated official web

0.31 2015/08/12
  - Updated internals for latest mojolicious usage.
  - Added FormValidatorLazy plugin for zero-configuration form validation.
  - Improved tests.
  - Improved documents.

0.30 2014/06/07
  - Updated depencencies.
  - Fixed a test for perl-5.8.18.
  - Started travis ci test.
  - Adopted scss and now csses are auto-generated.

0.29 2014/06/01
  - Adopted html5 for core templates and examples.

0.28 2014/02/11
  - Added support for rotating secrets.
  - Removed secret method from Marquee.
  - Now render auto detects path if not given.
  - Changed status code for directory traversal from 404 to 400.
  - Fixed bug on viax.
  - Updated mojo API usage in favor of v4.78.
  - Ported latest implementation of Mojolicious::Types.
  - Improved POD viewer a little.

0.27 2013/11/11
  - Now content type header is auto set when missing.
  - Now dynamic renderer do not overwrite existing response.
  - Now Router plugin is auto activated and routes can be accessible as an app
    attribute.

0.26 2013/11/10
  - Added FormValidatorLazy plugin.
  - Refactored static/dinamic dispach logic to serve method.
  - Changed router api for consistency.
  - Made unserved route not tot fallback to default route.
  - Added viax method to router API.
  - Improved MIME type detection.

0.25 2013/05/31
  - Added SCSS plugin
  - Updated API usage in favor of mojo v4.13

0.24 2013/05/31
  - Fixed a bug on Official site example.
  - Now it depends on Mojolicious v4.09.

0.23 2012/11/26
  - Improved installation direction.

0.22 2012/11/26
  - Improved documentations.
  - Added install.sh

0.21 2012/10/21
  - Removed deprecated methods.
  - Impoved router on capturing hehavior.

0.20 2012/10/01
  - Fixed that some tests have been unpassed.
  - Fixed potential bug on debug screen .

0.19 2012/09/28
  - Refactored add_handler.
  - Deprecated add_handler in favor of $app->dynamic->add_handler.
  - Improved debug screen.

0.18 2012/09/20
  - Refactored many features into classes.
  - Deprecated path_to_type in favor of $app->types->type_by_path.
  - Deprecated render_ssi in favor of $app->dynamic->render.
  - Deprecated search_static in favor of $app->static->search.
  - Deprecated search_template in favor of $app->dynamic->search.
  - Deprecated serve_static in favor of $app->static->serve.
  - Deprecated serve_dynamic in favor of $app->dynamic->serve.
  - Deprecated ssi_handlers in favor of $app->dynamic->handlers.

0.17 2012/09/14
  - Improved documentations.
  - Fixed ETag plugin to cause less undef warnings.
  - Changed API of Marquee::Cache not to accept multiple expiretion code.

0.16 2012/09/11
  - Added ETag plugin.
  - Added maxage option for static file caching and refactored to a class.
  - Fixed official site in Japanese.

0.15 2012/09/10
  - Improved bundle css styles.
  - Added session function for ep renderer.
  - Fixed some documents.

0.14 2012/09/07
  - Improved documentations.

0.13 2012/09/05
  - Improved documentations.
  - Improved plugin API to accept multiple arguments.
  - Added experimental plugin AuthPretty.

0.12 2012/09/03
  - Changed domain for official site.
  - Improved documentations.

0.11 2012/09/01
  - Fixed many bugs on windows.
  - Added cookbook.
  - Added paths options for pod viewer.
  - Added official site app.
  - Improved documentaions.

0.10 2012/08/30
  - Fixed a test

0.09 2012/08/30
  - Fixed bundling HTML to be valid.
  - Fixed debug screen for non-505 errors.
  - Improved documentation.

0.08 2012/08/29
  - Added markdown plugin and a command line option.
  - Added extends_as function for ep handler.
  - Added POD translations for Japanese.
  - Improved documentations.
  - Improved POD viewer and.
  - Improved context class API.
  - Changed router plugin API.
  - Changed that leading slash for to_abs indidates Marquee root.
  - Fixed a bug on default_file handling.
  - Fixed start method to catchup mojo Commit:2111684fe5b986d00ec5e8abef405d414aa34086

0.07 2012/08/04
  - Added include_as function to include a file as any SSI handlers.
  - Added warnings for template function redefinision.
  - Added EPHelperExample plugin.
  - Added a feature that include function got capable of static files.
  - Changed EP API that include and extends functions not to accept template
    extentions.
  - Added add_cond method for router plugin
  - Fixed some PODs.

0.06 2012/07/24
  - Added tree mode to auto index.
  - Renamed distribution from MojoSimpleHTTPServer to Marquee.
  - Improved iter function to accept blessed obj in array.
  - Removed basic auth from router in favor of Auth plugin.

0.05 2012/06/20
  - Added ability to manage session data by signed cookie.
  - Added pod viewer for commandline.

0.04 2012/06/05
  - Fixed a bug that aroud_dispatch hook has been invoked on wrong condition.
  - Added iter function for array iteration.
  - Added MSHS class for context access shortcut.
  - Added MIME types for auto index.

0.03 2012/06/04
  - Made error documents replaceable.
  - Improved design for error documents and auto index.

0.02 2012/06/02
  - Fixed directory traversal vulnerability.
  - Improved template cache to expire on template change.

lib/Marquee.pm
package Marquee;
use strict;
use warnings;
use Mojo::Base -base;
use feature 'signatures';
no warnings "experimental::signatures";
use File::Spec::Functions;
use Digest::MD5 qw(md5_hex);
use Carp;
use Mojo::Asset::File;
use Mojo::URL;
use Mojo::Home;
use Mojolicious::Commands;
use Mojo::Exception;
use Mojo::Log;
use Mojo::Transaction::HTTP;
use Mojo::Util;
use Marquee::Context;
use Marquee::Dynamic;
use Marquee::ErrorDocument;
use Marquee::Hooks;
use Marquee::SSIHandler::EP;
use Marquee::SSIHandler::EPL;
use Marquee::Stash;
use Marquee::Static;
use Marquee::Types;
our $VERSION = '0.42';

our $CONTEXT;

has 'document_root';
has 'default_file';
has dynamic => sub {Marquee::Dynamic->new};
has error_document => sub {Marquee::ErrorDocument->new};
has home => sub { Mojo::Home->new->detect(ref shift) };
has log              => sub {
  my $self = shift;

  # Check if we have a log directory that is writable
  my $log  = Mojo::Log->new;
  my $home = $self->home;
  my $mode = $self->mode;
  $log->path($home->child('log', "$mode.log"))
    if -d $home->child('log') && -w _;

  # Reduced log output outside of development mode
  return $mode eq 'development' ? $log : $log->level('info');
};
has hooks => sub {Marquee::Hooks->new};
has mode     => sub { $ENV{MOJO_MODE} || $ENV{PLACK_ENV} || 'development' };
has roots => sub {[]};
has route => sub {
    my $router = $_[0]->plugin(Router => sub {});
    return $router->route;
};
has secrets => sub {[md5_hex($^T. $$. rand(1000000))]};
has stash => sub {Marquee::Stash->new};
has static => sub {Marquee::Static->new};
has types => sub { Marquee::Types->new };
has under_development => 0;
has x_powered_by => 'Marquee(Perl)';

### --
### Constructor
### --
sub new($class, @args) {
    my $self = $class->SUPER::new(@args);
    
    ### hook points
    $self->hook(around_dispatch => sub($) {
        $CONTEXT->app->dispatch;
    });
    $self->hook(around_static => sub($, @args) {
        $CONTEXT->app->static->serve(@args);
    });
    $self->hook(around_dynamic => sub($, @args) {
        $CONTEXT->app->dynamic->serve(@args);
    });
    
    $self->dynamic->add_handler(
                    ep => Marquee::SSIHandler::EP->new(log => $self->log));
    $self->dynamic->add_handler(
                    epl => Marquee::SSIHandler::EPL->new(log => $self->log));
    
    return $self;
}

### ---
### Asset directory
### ---
sub asset($class, $name=undef) {
    my $pm = $class. '.pm';
    $pm =~ s{::}{/}g;
    my @seed = (substr($INC{$pm}, 0, -3), 'Asset');
    return catdir(@seed, $name) if ($name);
    return catdir(@seed);
}

sub build_tx {
  return Mojo::Transaction::HTTP->new;
}

sub config { Mojo::Util::_stash(config   => @_) }

### --
### Shortcut for context
### --
sub c($, $set=undef) {
    return $set ? $CONTEXT = $set : $CONTEXT;
}

### --
### Accessor for localized context
### --
sub context($, $set=undef) {
    return $set ? $CONTEXT = $set : $CONTEXT;
}

### --
### dispatch
### --
sub dispatch($self) {
    
    my $path = $CONTEXT->req->url->path->clone->canonicalize;
    
    if (! $CONTEXT->served) {
        if ($path =~ $self->dynamic->handler_re) {
            $self->error_document->serve(403);
            return;
        }
    }
    
    if (! $CONTEXT->served) {
        $self->serve();
    }
    
    if (! $CONTEXT->served) {
        if (! $path->trailing_slash && scalar @{$path->parts}
                                            && $self->is_directory($path)) {
            my $uri = $CONTEXT->req->url->clone->path(
                                $path->clone->trailing_slash(1))->to_abs;
            $self->serve_redirect($uri);
        }
    }
}

### --
### handler
### --
sub handler($self, $tx) {
    
    local $CONTEXT = Marquee::Context->new(app => $self, tx => $tx);
    
    $self->_init;
    
    $tx->res->headers->header('X-Powered-By' => $self->x_powered_by);

    eval {
        $self->hooks->emit_chain('around_dispatch');
    };
    
    if ($@) {
        $self->log->fatal("Processing request failed: $@");
        $self->error_document->serve(500, $self->under_development && $@);
    }
    
    if (! $CONTEXT->served) {
        $self->error_document->serve(404);
        $self->log->fatal($tx->req->url->path. qq{ Not found});
    }
    
    if ($tx->res->code == 200 && ! defined $tx->res->headers->content_type) {
        if (my $type = $self->types->type_by_path($tx->req->url->path)) {
            $tx->res->headers->content_type($type);
        }
    }
    
    $CONTEXT->close;
    
    $tx->resume;
}

### --
### Add hook
### --
sub hook($self, @args) {
    $self->hooks->on(@args);
}

### --
### Check if the path is a directory or not
### --
sub is_directory($self, $path) {
    for my $root (@{$self->roots}) {
        return 1 if (-d catdir($root, $path));
    }
}

### --
### Set log file path
### --
sub log_file($self, $path) {
    return $self->log->path($path);
}

### --
### Register plugin
### --
sub plugin($self, $name, @args) {
    
    unless ($name =~ s/^\+//) {
        $name = "Marquee::Plugin\::$name";
    }
    
    if (! $name->can('register')) {
        my $file = $name;
        $file =~ s!::!/!g;
        require "$file.pm"; ## no critic
    }
    my $plug = $name->new;
    $plug->register($self, @args);
    return $plug;
}

### --
### server
### mojo v7.80 (commit:57310d5)
### --
sub server { }

### --
### serve
### --
sub serve($self, $path=undef) {
    
    if (! defined $path) {
        $path = $CONTEXT->req->url->path->clone->canonicalize;
        if (@{$path->parts}[0] && @{$path->parts}[0] eq '..') {
            $self->error_document->serve(400);
            my $path = $CONTEXT->req->url->path;
            $self->log->fatal(qq{Bad request path "$path", possible directory traversal attempt.});
            return;
        }
        $path = _auto_fill_filename($path->clone, $self->default_file);
        $path->leading_slash(0);
    }
    
    if (my $try1 = $self->static->search($path)) {
        $self->hooks->emit_chain('around_static', $try1);
    } elsif (my $try2 = $self->dynamic->search($path)) {
        $self->hooks->emit_chain('around_dynamic', $try2);
    }
}

### --
### serve redirect
### --
sub serve_redirect($self, $uri) {
    $CONTEXT->res->code(301);
    $CONTEXT->res->headers->location($self->to_abs($uri)->to_string);
}

### --
### generate absolute uri
### --
sub to_abs($self, $url) {
    
    $url = Mojo::URL->new($url);
    
    if (! $url->scheme) {
        my $base = $CONTEXT->req->url->clone;
        $base->userinfo(undef);
        $url->base($base);
    }
    
    return $url->to_abs;
}

### --
### start application
### --
sub start($self, @args) {
    $self->_init;
    Mojolicious::Commands->new(app => $self)->run(@args ? @args : @ARGV);
}

### --
### auto fill files
### --
sub _auto_fill_filename($path, $default=undef) {
    if ($default && ($path->trailing_slash || ! @{$path->parts})) {
        push(@{$path->parts}, $default);
        $path->trailing_slash(0);
    }
    return $path;
}

### --
### init
### --
sub _init($self) {
    
    return if ($self->{_inited});
    
    $self->{_inited} = 1;
    
    if (! $self->document_root || ! -d $self->document_root) {
        die 'document_root is not a directory';
    }
    
    unshift(@{$self->roots}, canonpath($self->document_root), __PACKAGE__->asset);
    
    $self->static->roots($self->roots);
    $self->dynamic->roots($self->roots);
}

1;

__END__

=head1 NAME

Marquee - Yet another Mojo based web framework

=head1 SYNOPSIS

use Marquee directly.

    #!/usr/bin/env perl
    use strict;
    use warnings;
    
    use File::Basename 'dirname';
    use File::Spec;
    use lib join '/', File::Spec->splitdir(File::Spec->rel2abs(dirname(__FILE__))), 'lib';
    
    use Marquee;
    
    my $app = Marquee->new;
    $app->document_root($app->home->rel_dir('public_html'));
    $app->start;

Inherited application.

    package MyApp;
    use Mojo::Base 'Marquee';
    
    sub new($class, @args) {
        my $self = $class->SUPER::new(@args);
        
        $self->document_root($self->home->rel_dir('public_html'));
        $self->log_file($self->home->rel_dir('mojo_log/Marquee.log'));
        $self->default_file('index.html');
        $self->under_development(1);
        $self->secrets(['g3y3nheher']);
        
        return $self;
    }
    
    package main;
    
    MyApp->new->start;

=head1 DESCRIPTION

L<Marquee> distribution is yet another web framework built on mojo modules
in L<Mojolicious> distribution, designed to make dynamic content
development to be plotted at an extension of designer work.

=head1 ATTRIBUTES

L<Marquee> inherits all attributes from L<Mojo> and implements the following
new ones.

=head2 C<document_root>

Specify a path to document root directory. The directory can contain both static
files and templates.

    $app->document_root($app->home->rel_dir('public_html'));

=head2 C<default_file>

Specify a default file name and activate auto fill. The auto fill will occurs
when the request path is trailing slashed.

    $app->default_file('index.html');

=head2 C<dynamic>

L<Marquee::Dynamic> class instance.

    $app->dynamic(Marquee::Dynamic->new);
    my $dynamic = $app->dynamic;

=head2 C<error_document>

Error document renderer instance. Defaults to L<Marquee::ErrorDocument>.

    $app->error_document(Marquee::ErrorDocument->new);

=head2 C<hooks>

A L<Marquee::Hooks> instance.

    $app->hooks(Marquee::Hooks->new);

=head2 C<roots>

Array of paths that contains static and templates. Marquee tries to find asset
files in ascend order in the array. The array is started with L</document_root>
copy, and followed by bundle directories for Marquee core and plugins.

    push(@{$app->roots}, 'path/to/additional_dir')

=head2 C<secrets>

A secret passphrases used for signed cookies and the like, defaults to random
string. By changing this, you can expire all signed cookies at once.

    my $secrets = $app->secrets;
    $app       = $app->secrets(['passw0rd']);

Only the first passphrase is used to create new signatures, but all of them for
verification. So you can increase security without invalidating all your
existing signed cookies by rotating passphrases, just add new ones to the
front and remove old ones from the back.

  # Rotate passphrases
  $app->secrets(['new_passw0rd', 'old_passw0rd', 'very_old_passw0rd']);

=head2 C<stash>

An L<Marquee::Stash> instance. Though Marquee's stash is localized and cloned
per request, this also can contain persistent values for application specific
and can be referred transparently from anywhere.

    $app->stash(Marquee::Stash->new);
    my $stash = $app->stash;

=head2 C<static>

L<Marquee::Static> class instance.

    $app->static(Marquee::Static->new);
    my $static = $app->static;

=head2 C<types>

Contains L<Mojolicious::Type> instance.

    my $type = $app->types;
    $type->type(zip => 'application/zip');

=head2 C<under_development>

Activate debug screen, defaults to undef.

    $app->under_development(1);

=head2 C<x_powered_by>

Set X-POWERED-BY response header.

    $app->x_powered_by('MyApp');

The header appears as follows.

    Content-Type: text/html;charset=UTF-8
    X-Powered-By: MyApp
    Server: Mojolicious (Perl)

=head1 CLASS METHODS

L<Marquee> inherits all class methods from L<Mojo> and implements the following
new ones.

=head2 new

Constructor.

    my $app = Marquee->new;

=head2 C<asset>

Returns bundled asset path for given file name. If the file name not specified,
The method returns the asset directory.

    my $asset = Marquee->asset('path/to/common.css');
    
    say $asset # /path/to/lib/Marquee/Asset/path/to/common.css
    
    my $asset = Marquee->asset();
    
    say $asset # /path/to/lib/Marquee/Asset

This method allows you to bundle files for perl modules in separated name spaces.
The following is an example for getting bundle files of arbitrary module.

    my $asset = SomePackage->Marquee::asset('path/to/common.css');
    
    say $asset # /path/to/lib/SomePackage/Asset/path/to/common.css
    
    my $asset = SomePackage->Marquee::asset();
    
    say $asset # /path/to/lib/SomePackage/Asset

=head1 INSTANCE METHODS

L<Marquee> inherits all instance methods from L<Mojo> and implements the
following new ones.

=head2 config

    my $hash = $app->config;
    my $foo  = $app->config('foo');
    $app     = $app->config({foo => 'bar', baz => 23});
    $app     = $app->config(foo => 'bar', baz => 23);

Application configuration.

    # Remove value
    my $foo = delete $app->config->{foo};
    
    # Assign multiple values at once
    $app->config(foo => 'test', bar => 23);

=head2 C<c>

An alias for L</context> method.

=head2 C<context>

Returns current context. This refers to C<$Marquee::CONTEXT> localized
per request.

    my $c = $app->context;

=head2 C<dispatch>

Front dispatcher.

    $app->dispatch()

=head2 C<handler>

Handler called by mojo layer. This is the base point for every request and sets
a response to C<$tx>.

    $app->handler($tx)

=head2 C<hook>

Alias to $app->hooks->on. This adds a callback for specified hook point.
These hooks are currently available:

=over 2

=item C<around_dispatch>

Wraps dispatch process.

    $app->hook(around_dispatch => sub($next) {
        ### pre-process
        $next->();
        ### post-process
    });

=item C<around_static>

Wraps static dispatch process.

    $app->hook(around_static => sub($next, @args) {
        ### pre-process
        $next->(@args);
        ### post-process
    });

=item C<around_dynamic>

Wraps dynamic dispatch process.

    $app->hook(around_dynamic => sub($next, @args) {
        ### pre-process
        $next->(@args);
        ### post-process
    });

=back

=head2 C<is_directory>

Returns if the path is directory. The search is made against the directories in
L</roots> attribute paths.

    $app->is_directory('/path/to/directory') # bool

=head2 C<log_file>

Set log file

    $app->log_file('/path/to/file')

=head2 C<plugin>

Load a class as a plugin. The prefix L<Marquee::Plugin> is prepended unless the
class name C<$class> begins with C<+> sign, which means the class name is
already fully qualified.

    my $plugin = $app->plugin(MyPlug => @params); # Marquee::Plugin::MyPlug
    my $plugin = $app->plugin('+MyPlugins::MyPlug' => @params); # MyPlugins::MyPlug

=head2 C<serve>

Serves specific content for given file path. This internally searches static
contents and dynamic contents.

    $app->serve('index.html');

=head2 C<serve_redirect>

Serves response that redirects to given URI.

    $app->serve_redirect('http://example.com/');
    $app->serve_redirect('/path/');

=head2 C<start>

Starts app

    $app->start();

=head2 C<to_abs>

Generates absolute URI for given path along to the request URI.

On request to https://example.com:3001/a/index.html
The example below generates https://example.com:3001/path/to/file.html
    
    say $self->to_abs('/path/to/file.html');

=head1 SEE ALSO

L<Mojolicious>

=head1 AUTHOR

Sugama Keita, E<lt>sugama@jamadam.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2012 by Sugama Keita.

This program is free software; you can redistribute it and/or
modify it under the same terms as Perl itself.

=cut
